jupyter-extension/jupyterlab_pachyderm/pps_client.py

```diff
@@ -0,0 +1,79 @@
+import os.path
+import json
+import subprocess
+from tempfile import NamedTemporaryFile
+
+
+from tornado.httpclient import AsyncHTTPClient
+
+from .log import get_logger
+
+
+class PPSClient:
+    """Client interface for the PPS extension backend."""
+
+    def __init__(self):
+        self.client = AsyncHTTPClient()
+
+    async def generate(self, path, config):
+        """Generates the pipeline spec from the Notebook file specified.
+
+        Args:
+            path: The path (within Jupyter) to the Notebook file.
+            config: The PPS configuration for the Notebook file.
+        """
+        get_logger().debug(f"path: {path} | body: {config}")
+        # script = await self.client.fetch(f"http://localhost:8888/nbconvert/python{path}")
+        script_name = os.path.basename(path)
+        pipeline_spec = create_pipeline_spec("test_pipeline", "python:3.10", dict(pfs=dict(repo="data")), script_name)
+        return json.dumps(pipeline_spec)
+
+    async def create(self, path, config):
+        """Creates the pipeline from the Notebook file specified.
+
+        Args:
+            path: The path (within Jupyter) to the Notebook file.
+            config: The PPS configuration for the Notebook file.
+        """
+        get_logger().debug(f"path: {path} | body: {config}")
+        input_spec = config.get("input")
+        with NamedTemporaryFile() as temp_config:
+            # TODO: Write the SAME config as a yaml file
+            subprocess.run(
+                ["same", "run", "--same-file", temp_config.name, "--target", "pachyderm", "--input", str(input_spec)]
+            )
+        return json.dumps(dict())
+
+
+def create_pipeline_spec(
```

Comment:
Doesn't `SAME` generate this? (Fine to copy it over, just curious)

Reply:
It does generate this, there's just no current way to get SAME to return it to us without creating the pipeline. I figured this would be easier to debug the extension with in the interim